shell/platform/android/io/flutter/embedding/engine/renderer/FlutterRenderer.java

```diff
@@ -360,5 +382,104 @@ public void dispatchSemanticsAction(
     public int systemGestureInsetRight = 0;
     public int systemGestureInsetBottom = 0;
     public int systemGestureInsetLeft = 0;
+    public List<DisplayFeature> displayFeatures = new ArrayList<DisplayFeature>();
+  }
+
+  /**
+   * Area of the viewport obstructed by a feature, eg. a hinge, a fold crease or camera cutout
+   *
+   * <p>Similar to {@link androidx.window.DisplayFeature}, used for reporting Fold / Hinge areas.
+   * The reason we define our own model is that we also want to support cutouts.
+   */
+  public static final class DisplayFeature {
+    public final Rect bounds;
+    public final DisplayFeatureType type;
+    public final DisplayFeatureState state;
+
+    public DisplayFeature(Rect bounds, DisplayFeatureType type, DisplayFeatureState state) {
+      this.bounds = bounds;
+      this.type = type;
+      this.state = state;
+    }
+
+    public DisplayFeature(Rect bounds, DisplayFeatureType type) {
+      this.bounds = bounds;
+      this.type = type;
+      this.state = DisplayFeatureState.UNKNOWN;
+    }
+  }
+
+  /**
+   * Types of display features that can obstruct the viewport.
+   *
+   * <p>Some, like FOLD, can be reported without actually impeding drawing on the screen. They are
+   * useful for knowing where the display is bent or has a crease. The {@link DisplayFeature} bounds
+   * can be 0-width in such cases.
+   */
+  public enum DisplayFeatureType {
+    /**
+     * We do not know this type of display feature yet. This can happen if WindowManager is updated
+     * with new types.
+     */
+    UNKNOWN(0),
+
+    /**
+     * A fold in the flexible screen without a physical gap. Corresponds to {@link
+     * androidx.window.DisplayFeature.TYPE_FOLD}
+     */
+    FOLD(1),
+
+    /**
+     * A physical separation with a hinge that allows two display panels to fold. Corresponds to
+     * {@link androidx.window.DisplayFeature.TYPE_HINGE}
+     */
+    HINGE(2),
+
+    /**
+     * A non-functional area of the screen, usually housing cameras or sensors. Corresponds to
+     * {@link android.view.DisplayCutout}
+     */
+    CUTOUT(3);
+
+    public final int encodedValue;
+
+    DisplayFeatureType(int encodedValue) {
+      this.encodedValue = encodedValue;
+    }
+  }
+
+  /**
+   * State of the display feature.
+   *
+   * <p>For foldables, the state is the posture. For cutouts, this is {@link UNKNOWN}
+   */
+  public enum DisplayFeatureState {
+    /** The display feature is a cutout or this state is new and not yet known to Flutter. */
```

Comment:
Is it a good idea to combine the "new and not yet known to flutter" with the "feature is a cutout" since the cutout one is invalid while the other is unknown.

Reply:
I see your point. I am ok with this, since checking if a display feature is a cutout or not is meant to be checked through `type`, not `state`. The comment about cutouts here is meant as extra information about how `state` is populated.